sqlx-core/src/sqlite/types/chrono.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{
+    decode::Decode,
+    encode::{Encode, IsNull},
+    error::BoxDynError,
+    sqlite::{type_info::DataType, Sqlite, SqliteArgumentValue, SqliteTypeInfo, SqliteValueRef},
+    types::Type,
+};
+use chrono::prelude::*;
+
+impl Type<Sqlite> for NaiveDateTime {
+    fn type_info() -> SqliteTypeInfo {
+        SqliteTypeInfo(DataType::Datetime)
+    }
+}
+
+impl Encode<'_, Sqlite> for NaiveDateTime {
+    fn encode_by_ref(&self, buf: &mut Vec<SqliteArgumentValue<'_>>) -> IsNull {
+        let text: String = self.format("%F %T%.f").to_string();
+        Encode::<Sqlite>::encode(text, buf)
+    }
+}
+
+impl<'a> Decode<'a, Sqlite> for NaiveDateTime {
+    fn decode(value: SqliteValueRef<'a>) -> Result<Self, BoxDynError> {
+        decode_naive_from_text(value.text()?)
```

Comment:
Can you match on the type and handle `Int`, `Int64`, and `Float` storage formats as well?

```rust
match value.type_info().0 {
  DataType::Int => ...
}
```

See: https://www.sqlite.org/datatype3.html#date_and_time_datatype

> TEXT as ISO8601 strings ("YYYY-MM-DD HH:MM:SS.SSS").

> REAL as Julian day numbers, the number of days since noon in Greenwich on November 24, 4714 B.C. according to the proleptic Gregorian calendar.

> INTEGER as Unix Time, the number of seconds since 1970-01-01 00:00:00 UTC.

Reply:
@mehcode Thanks for the review!  I've tried to match the `type_info()`, but it seems that I always get `None`.  I've put my [WIP on a commit 3e507eb8](https://github.com/launchbadge/sqlx/commit/3e507eb8ac30955523427f1d8a4b56ce383c271e), but the gist of it is that I'm doing:

```rust
    fn decode(value: SqliteValueRef<'a>) -> Result<Self, BoxDynError> {
        println!("YYY: {:?}", value.type_info());
        if let Some(type_info) = value.type_info() {
```

and

```rust
    test_decode_type!(chrono_date_time_utc_decode<DateTime::<Utc>>(Sqlite,
        "'1996-12-20 00:39:57.123'" == Utc.ymd(1996, 12, 20).and_hms_milli(0, 39, 57, 123),
        "CAST(2459028.37981481 AS FLOAT)" == Utc.ymd(2020, 06, 27).and_hms(21, 06, 56),
```

But even with the explicit `CAST(... AS FLOAT)`, when running `cargo test --features sqlite,chrono -- --nocapture` I still get:

```
...
YYY: None
test chrono::test_prepared_type_chrono_date_time_local ... ok
Error: error occurred while decoding column test chrono::test_unprepared_type_chrono_naive_date_time ... ok
0: encountered unexpected or invalid data: Did not find a matching pattern

Caused by:
    encountered unexpected or invalid data: Did not find a matching pattern
thread 'chrono::test_unprepared_type_chrono_date_time_utc_decode' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /Users/felipe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libstd/macros.rs:16:9
...
```

I'm thinking that maybe I could try the numeric versions whenever `.text()` fails, but I'm not sure how I'd be able to choose between Unix timestamps and Julian.  Or maybe there's another way I'm missing?

Reply:
Yeah... This comment actually reminded me we had a pretty big hole in our type logic for SQLite. That's all fixed now which enables the type switch during decoding. Thank you.